File: rust/src/circuit.rs
use std::cmp::Ordering;
use super::components::component::Component;
use std::collections::BTreeSet;
use std::sync::{Arc, MutexGuard};
use std::sync::Mutex;
use crate::util::ArrayBuilder;

static CIRCUIT_INDEX_COUNTER: Mutex<u64> = Mutex::new(0);

pub struct Circuit {
    connection_points: Mutex<BTreeSet<Arc<ConnectionPoint>>>,
    components: Mutex<Vec<Box<dyn Component>>>,
    has_ground: bool,
    connection_point_counter: Mutex<u64>,
    id: u64,
}

impl Circuit {
    pub fn new() -> Circuit {
        let mut counter: MutexGuard<u64> = CIRCUIT_INDEX_COUNTER.lock().unwrap();

        let ground: ConnectionPoint = ConnectionPoint {
            id: 0,
            voltage: 0.0,
            circuit_id: *counter,
        };

        let circuit: Circuit = Circuit {
            connection_point_counter: Mutex::new(0),
            components: Mutex::new(Vec::new()),
            connection_points: Mutex::new(BTreeSet::new()),
            has_ground: true,
            id: *counter,
        };

        *counter += 1;

        circuit.connection_points.lock().unwrap().insert(Arc::new(ground));

        circuit
    }

    pub fn new_connection_point(&mut self) -> Arc<ConnectionPoint> {
        let mut connection_points = self.connection_points.lock().unwrap();
        let point: Arc<ConnectionPoint> = Arc::new(ConnectionPoint::new(self));
        connection_points
            .insert(point.clone());
        point
    }

    pub fn get_ground(&self) -> Arc<ConnectionPoint> {
        //TODO maybe make this better in the future?
        assert!(self.has_ground);
        self.connection_points.lock().unwrap().first().unwrap().clone()
    }

    pub fn add_component(& mut self, component: impl Component + 'static)
    {
        self.components.lock().unwrap().push(Box::new(component));
    }

    pub fn simulation_step(&mut self, dt: f64) {

        let connection_points = self.connection_points.lock().unwrap();
        let components = self.components.lock().unwrap();

        let num_voltages = connection_points.len();
        let num_currents: usize = components.iter().map(|c| c.connection_count()).sum();
        let num_to_solve_for: usize = num_voltages + num_currents;

        let mut matrix: Vec<Vec<f64>> = Vec::with_capacity(num_to_solve_for);

        let mut constants: Vec<f64> = Vec::with_capacity(num_to_solve_for);


        let mut component_connection_index: usize = 0;

        //Enter in equations for each component
        for component in &*components {

            let connections: Box<[(Arc<ConnectionPoint>, Arc<ConnectionPoint>, &dyn Component)]> = component.connections();
            let constraints: Box<[f64]> = component.constraints();
            let mut current_connection: usize = 0;

            for i in 0..connections.len() {

                let mut coefficients: Vec<f64> = Vec::with_capacity(num_to_solve_for);

                let current_dependence_index: usize = (2 * component.connection_count() + 1) * i;
                let voltage_dependence_index: usize = current_dependence_index + component.connection_count();
                let constant_dependence: f64 = constraints[2 * component.connection_count() + 1 * (2 * component.connection_count() + 1)];

                for j in 0..connections.len() {
                    let connection = &connections[j];

                    coefficients[(component_connection_index - current_connection) + j] = constraints[current_dependence_index + j];

                    coefficients[num_currents + self.connection_point_position(first_connection_point(connection))] += constraints[voltage_dependence_index + j];
                    coefficients[num_currents + self.connection_point_position(second_connection_point(connection))] += -constraints[voltage_dependence_index + j]
                }

                constants[component_connection_index] = constant_dependence;
                matrix[component_connection_index] = coefficients;

                component_connection_index += 1;
                current_connection += 1;
            }
        }

        // values for kirchhoff's junction laws
        let mut node_index: usize = num_currents;
        for connection_point in &*self.connection_points.lock().unwrap() {

            let mut coefficients: Vec<f64> = Vec::with_capacity(num_to_solve_for);

            if connection_point.eq(&self.get_ground()) {
                //Ground node should always be at 0
                coefficients[num_currents] = 1.0;
            }
            else {
                let mut arr: ArrayBuilder<(Arc<ConnectionPoint>, Arc<ConnectionPoint>, &dyn Component)> = ArrayBuilder::new();
                components.iter().for_each(|c|  arr.push_array(c.connections()));
                let allConnections

            }

        }

    }

    #[inline]
    fn connection_point_position(&self, point: &Arc<ConnectionPoint>) -> usize {
        self.connection_points.lock().unwrap().iter().position(|p| (**p).eq(&**point)).unwrap()
    }
}

#[inline]
fn first_connection_point<'a>(pair: &'a(Arc<ConnectionPoint>, Arc<ConnectionPoint>, &'a dyn Component)) -> &'a Arc<ConnectionPoint> {
    let (first, second, _) = pair;
    match first.cmp(second) {
        Ordering::Less => first,
        Ordering::Equal => unreachable!(), //Might not actually be unreachable if components create garbage conenctions
        Ordering::Greater => second,
    }
}


#[inline]
fn second_connection_point<'a>(pair: &'a(Arc<ConnectionPoint>, Arc<ConnectionPoint>, &'a dyn Component)) -> &'a Arc<ConnectionPoint> {
    let (first, second, _) = pair;
    match first.cmp(second) {
        Ordering::Less => second,
        Ordering::Equal => unreachable!(), //Might not actually be unreachable if components create garbage conenctions
        Ordering::Greater => first,
    }
}



pub struct ConnectionPoint {
    id: u64,
    circuit_id: u64,
    voltage: f64,
}

impl ConnectionPoint {
    fn new(circuit: &Circuit) -> ConnectionPoint {
        *circuit.connection_point_counter.lock().unwrap() += 1;
        ConnectionPoint {
            id: *circuit.connection_point_counter.lock().unwrap(),
            voltage: 0.0,
            circuit_id: circuit.id,
        }
    }

    #[inline]
    fn set_voltage(&mut self, voltage: f64) {
        self.voltage = voltage;
    }

    #[inline]
    pub fn get_voltage(&self) -> f64 {
        self.voltage
    }
}

impl PartialEq for ConnectionPoint {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
            && self.circuit_id
                == other.circuit_id
    }
}

impl Eq for ConnectionPoint {}

impl PartialOrd for ConnectionPoint {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        match self.circuit_id.cmp(&other.circuit_id) {
            Ordering::Equal => Some(self.id.cmp(&other.id)),
            Ordering::Greater => Some(Ordering::Greater),
            Ordering::Less => Some(Ordering::Less),
        }
    }
}

impl Ord for ConnectionPoint {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap()
    }
}

//Designed to be thread safe, since simulation may want to be deferred to another thread
unsafe impl Send for Circuit {}
unsafe impl Sync for Circuit {}

unsafe impl Send for ConnectionPoint {}
unsafe impl Sync for ConnectionPoint {}

File: java/lib/src/main/java/xyz/kpzip/kspice/circuit/Circuit.java
package xyz.kpzip.kspice.circuit;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Objects;
import java.util.TreeSet;

import org.apache.commons.math3.linear.Array2DRowRealMatrix;
import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.DecompositionSolver;
import org.apache.commons.math3.linear.LUDecomposition;
import org.apache.commons.math3.linear.RealMatrix;
import org.apache.commons.math3.linear.RealVector;

import xyz.kpzip.kspice.components.Component;
import xyz.kpzip.kspice.util.ArrayBuilder;

/**
 * Represents an electronic circuit.
 * Components can be added with addComponent(Component c), and new connection points can be made with createConnectionPoint().
 * To simulate, call simulationStep(double dt)
 * 
 * @author kpzip
 *
 */
public sealed class Circuit permits Subcircuit {
	
	public static boolean DEBUG = false;
	
	//Nodes should always be in ascending order with no gaps e.g. 0,1,2,3,4,5 NOT 0,2,3,5,6,7,9
	//Positive Current is determined to flow from high id nodes to low id nodes
	protected TreeSet<ConnectionPoint> connectionPoints;
	protected ArrayList<Component> components;
	private ConnectionPoint ground;
	private int connectionPointIndex = 1;
	
	public Circuit() {
		this(true);
	}
	
	//Constructor to allow 
	protected Circuit(boolean gnd) {
		connectionPoints = new TreeSet<ConnectionPoint>();
		components = new ArrayList<Component>();
		if (gnd) {
			ground = new Circuit.ConnectionPoint(0);
			connectionPoints.add(ground);
		}
		else {
			ground = null;
		}
	}
	
	public void simulationStep(double dt) {
		
		int numVoltages = connectionPoints.size();
		int numCurrents = components.stream().mapToInt((c) -> c.connectionCount()).sum();
		int numToSolveFor = numVoltages + numCurrents;
		
		//The matrix representing the coefficients in the system
		double [] [] matrix = new double[numToSolveFor] [numToSolveFor];
		
		//The vector representing the vector (The constants in each equation)
		double [] constants = new double[numToSolveFor];
		
		//enter values for component constraints
		int componentConnectionIndex = 0;

		for (Component c : components) {
			
			ConnectionPointPair[] connections = c.connections();
			double[] constraints = c.constraints();
			
			for (int i = 0; i < connections.length; i++) {
				
				double[] coefficients = new double[numToSolveFor];
				
				int currentDependenceIndex = (2 * c.connectionCount() + 1) * i;
				int voltageDependenceIndex = currentDependenceIndex + c.connectionCount();
				double constantDependence = constraints[3*i + 2 * c.connectionCount()];
				
				//Add all of the constraints to the equation
				for (int j = 0; j < connections.length; j++) {
					ConnectionPointPair connection = connections[j];
					
					coefficients[componentConnectionIndex + j] = constraints[currentDependenceIndex + j];
					coefficients[numCurrents + connection.first.id] += constraints[voltageDependenceIndex + j];
					coefficients[numCurrents + connection.second.id] += -constraints[voltageDependenceIndex + j];
				}
				
				constants[componentConnectionIndex] = constantDependence;
				matrix[componentConnectionIndex] = coefficients;
				
				componentConnectionIndex++;
			}
		}
		
		//enter values for Kirchhoff's junction laws
		int nodeIndex = numCurrents;
		for (Circuit.ConnectionPoint p : connectionPoints) {
			
			double[] coefficients = new double[numToSolveFor];
			
			if (p == this.ground) {
				//add equation that says ground node is at zero voltage
				coefficients[numCurrents] = 1;
			}
			else {
				ArrayBuilder<ConnectionPointPair> arr = new ArrayBuilder<ConnectionPointPair>(ConnectionPointPair.class, numCurrents);
				components.forEach((c) -> arr.pushArray(c.connections()));
			
				//ordered array of all connections in the circuit
				ConnectionPointPair[] allConnections = arr.toArray();
			
				//make it so all of the currents flowing into the node sum up to zero
				//Note: current flows from high node values to low node values
				for (int i = 0; i < allConnections.length; i++) {
					if (allConnections[i].first == p) {
						coefficients[i] = -1;
						continue;
					}
					if (allConnections[i].second == p) {
						coefficients[i] = 1;
					}
				}
			}
			matrix[nodeIndex] = coefficients;
			nodeIndex++;
			
		}
		
		//Debug: print out matrix and vector
		if (DEBUG) {
			System.out.println(Arrays.deepToString(matrix));
			System.out.println(Arrays.toString(constants));
		}
		
		//Now the matrix is populated, time to solve!
		RealMatrix A = new Array2DRowRealMatrix(matrix, false);
		RealVector B = new ArrayRealVector(constants, false);
		DecompositionSolver solver = new LUDecomposition(A).getSolver();
		RealVector X = solver.solve(B);
		double[] solutions = X.toArray();
		
		//set current and voltage values
		int valPtr = 0;
		for (Component c : components) {
			c.updateCurrent(Arrays.copyOfRange(solutions, valPtr, valPtr += c.connectionCount()));
		}
		
		for (ConnectionPoint p : connectionPoints) {
			p.voltage = solutions[valPtr];
			valPtr++;
		}
		components.forEach((c) -> c.differential(dt));
	}
	
	public void addComponent(Component c) {
		components.add(c);
	}
	
	public Circuit.ConnectionPoint createConnectionPoint() {
		ConnectionPoint c = new Circuit.ConnectionPoint();
		connectionPoints.add(c);
		return c;
	}
	
	//May return null
	public ConnectionPoint getGround() {
		return ground;
	}
	
	public void reset() {
		connectionPoints.forEach((p) -> p.reset());
		components.forEach((c) -> c.reset());
	}
	
	@Override
	public String toString() {
		StringBuilder str = new StringBuilder();
		for (Circuit.ConnectionPoint c : connectionPoints) {
			str.append(c.toString());
			str.append("\n");
		}
		return str.toString();
	}
	
	
	
	@Override
	public int hashCode() {
		return Objects.hash(components, connectionPoints);
	}

	public void validateConnectionPoints() {
		int connectionPointCheckerIndex = -1;
		for (Circuit.ConnectionPoint c : connectionPoints) {
			if (c.id != ++connectionPointCheckerIndex) {
				c.id = connectionPointCheckerIndex;
			}
		}
		connectionPointIndex = ++connectionPointCheckerIndex;
	}
	
	public ArrayList<Component> getComponents() {
		return components;
	}
	
	public TreeSet<ConnectionPoint> getConnectionPoints() {
		return connectionPoints;
	}
	
	public class ConnectionPoint implements Comparable<ConnectionPoint> {
		
		//id 0 is ground, negative is not part of the circuit
		private int id;
		
		private double voltage = 0;
		
		//Only used for creating the ground node
		protected ConnectionPoint(int id) {
			this.id = id;
		}
		
		protected ConnectionPoint() {
			this.id = connectionPointIndex++;
		}
		
		public double getVoltage() {
			return voltage;
		}

		public void setVoltage(double voltage) {
			this.voltage = voltage;
		}
		
		public int getId() {
			return id;
		}

		@Override
		public int compareTo(ConnectionPoint o) {
			return this.id - o.id;
		}
		
		
		
		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + getEnclosingInstance().hashCode();
			result = prime * result + Objects.hash(id);
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj)
				return true;
			if (obj == null)
				return false;
			if (getClass() != obj.getClass())
				return false;
			ConnectionPoint other = (ConnectionPoint) obj;
			if (!getEnclosingInstance().equals(other.getEnclosingInstance()))
				return false;
			return id == other.id;
		}

		public String toString() {
			return "Node: " + id + " with voltage: " + voltage + "V" + (id == 0 ? " (GND)" : "");
		}

		private Circuit getEnclosingInstance() {
			return Circuit.this;
		}
		
		public void reset() {
			this.voltage = 0;
		}

	}

}
